File: packages/adapter-typescript/src/merger.ts
/**
 * Internal merge & precedence rules of the adapter — 1:1 port of
 * dart/ductus/lib/src/adapter/merger.dart.
 *
 * `annotation` overrides `derived` field by field; two manual sources with
 * different values for the same field are an error (fail fast) that cites
 * both sources.
 */

import {
  AdapterException,
  refToString,
  SourceKind,
  type GraphEdge,
  type GraphFlow,
  type GraphNode,
  type SourceRef,
} from './graph-model.js';

export interface MergeResult {
  nodes: GraphNode[];
  edges: GraphEdge[];
  flows: GraphFlow[];
}

function byRef(a: SourceRef, b: SourceRef): number {
  const byFile = a.file < b.file ? -1 : a.file > b.file ? 1 : 0;
  return byFile !== 0 ? byFile : a.line - b.line;
}

/** Stable sort by (file, line) (Array.prototype.sort is stable). */
function sortedByRef<T>(items: readonly T[], ref: (item: T) => SourceRef): T[] {
  return [...items].sort((a, b) => byRef(ref(a), ref(b)));
}

function equalValues(a: unknown, b: unknown): boolean {
  if (Array.isArray(a) && Array.isArray(b)) {
    return a.length === b.length && a.every((value, i) => value === b[i]);
  }
  return a === b;
}

/** Conflict value format as in the Dart adapter: lists as `[a, b]`. */
function formatValue(value: unknown): string {
  return Array.isArray(value) ? `[${value.join(', ')}]` : String(value);
}

/**
 * Field-wise merge of a candidate list: the first manual value wins,
 * otherwise the first derived one; two different manual values ⇒ conflict.
 */
class FieldMerger<T> {
  readonly manual: T[];
  readonly derived: T[];

  constructor(
    private readonly kind: string,
    private readonly id: string,
    candidates: readonly T[],
    sourceOf: (candidate: T) => string,
    private readonly conflicts: string[],
  ) {
    this.manual = candidates.filter((c) => sourceOf(c) === SourceKind.annotation);
    this.derived = candidates.filter((c) => sourceOf(c) !== SourceKind.annotation);
  }

  merge<V>(
    field: string,
    get: (candidate: T) => V | undefined,
    ref: (candidate: T) => SourceRef,
  ): V | undefined {
    let value: V | undefined;
    let valueSource: T | undefined;
    for (const candidate of this.manual) {
      const v = get(candidate);
      if (v === undefined) continue;
      if (value === undefined) {
        value = v;
        valueSource = candidate;
      } else if (!equalValues(value, v)) {
        this.conflicts.push(
          `Conflict: ${this.kind} "${this.id}", field "${field}": ` +
            `"${formatValue(value)}" (${refToString(ref(valueSource as T))}) vs. "${formatValue(v)}" (${refToString(ref(candidate))}).`,
        );
      }
    }
    if (value !== undefined) return value;
    for (const candidate of this.derived) {
      const v = get(candidate);
      if (v !== undefined) return v;
    }
    return undefined;
  }
}

function mergeNodes(nodes: readonly GraphNode[], conflicts: string[]): GraphNode[] {
  const byId = new Map<string, GraphNode[]>();
  for (const node of sortedByRef(nodes, (n) => n.sourceRef)) {
    const group = byId.get(node.id);
    if (group === undefined) {
      byId.set(node.id, [node]);
    } else {
      group.push(node);
    }
  }

  const merged: GraphNode[] = [];
  for (const [id, candidates] of byId) {
    if (candidates.length === 1) {
      merged.push(candidates[0]!);
      continue;
    }
    const m = new FieldMerger('Node', id, candidates, (n) => n.source, conflicts);
    const ref = (n: GraphNode): SourceRef => n.sourceRef;
    const winner = m.manual.length > 0 ? m.manual[0]! : m.derived[0]!;
    const title = m.merge('title', (n) => n.title, ref);
    const flow = m.merge('flow', (n) => n.flow, ref);
    const description = m.merge('description', (n) => n.description, ref);
    merged.push({
      id,
      type: m.merge('type', (n) => n.type, ref) ?? winner.type,
      ...(title !== undefined ? { title } : {}),
      ...(flow !== undefined ? { flow } : {}),
      ...(description !== undefined ? { description } : {}),
      tags: m.merge('tags', (n) => (n.tags.length === 0 ? undefined : n.tags), ref) ?? [],
      source: winner.source,
      sourceRef: winner.sourceRef,
    });
  }
  return merged;
}

function mergeFlows(flows: readonly GraphFlow[], conflicts: string[]): GraphFlow[] {
  const byId = new Map<string, GraphFlow[]>();
  for (const flow of sortedByRef(flows, (f) => f.sourceRef)) {
    const group = byId.get(flow.id);
    if (group === undefined) {
      byId.set(flow.id, [flow]);
    } else {
      group.push(flow);
    }
  }

  const merged: GraphFlow[] = [];
  for (const [id, candidates] of byId) {
    if (candidates.length === 1) {
      merged.push(candidates[0]!);
      continue;
    }
    const m = new FieldMerger('Flow', id, candidates, (f) => f.source, conflicts);
    const ref = (f: GraphFlow): SourceRef => f.sourceRef;
    const winner = m.manual.length > 0 ? m.manual[0]! : m.derived[0]!;
    const title = m.merge('title', (f) => f.title, ref);
    const start = m.merge('start', (f) => f.start, ref);
    const description = m.merge('description', (f) => f.description, ref);
    merged.push({
      id,
      ...(title !== undefined ? { title } : {}),
      ...(start !== undefined ? { start } : {}),
      ...(description !== undefined ? { description } : {}),
      source: winner.source,
      sourceRef: winner.sourceRef,
    });
  }
  return merged;
}

function mergeEdges(edges: readonly GraphEdge[], conflicts: string[]): GraphEdge[] {
  const sorted = sortedByRef(edges, (e) => e.sourceRef);
  const manual = sorted.filter((e) => e.source === SourceKind.annotation);
  const derived = sorted.filter((e) => e.source !== SourceKind.annotation);

  // Manual edges with an explicit id: identity via id, field-wise merge.
  const result: GraphEdge[] = [];
  const manualById = new Map<string, GraphEdge[]>();
  for (const edge of manual) {
    if (edge.id === undefined) {
      result.push(edge);
    } else {
      const group = manualById.get(edge.id);
      if (group === undefined) {
        manualById.set(edge.id, [edge]);
      } else {
        group.push(edge);
      }
    }
  }
  for (const [id, candidates] of manualById) {
    if (candidates.length === 1) {
      result.push(candidates[0]!);
      continue;
    }
    const m = new FieldMerger('Edge', id, candidates, (e) => e.source, conflicts);
    const ref = (e: GraphEdge): SourceRef => e.sourceRef;
    const winner = candidates[0]!;
    const trigger = m.merge('trigger', (e) => e.trigger, ref);
    const label = m.merge('label', (e) => e.label, ref);
    const condition = m.merge('condition', (e) => e.condition, ref);
    result.push({
      id,
      from: m.merge('from', (e) => e.from, ref) ?? winner.from,
      to: m.merge('to', (e) => e.to, ref) ?? winner.to,
      ...(trigger !== undefined ? { trigger } : {}),
      ...(label !== undefined ? { label } : {}),
      ...(condition !== undefined ? { condition } : {}),
      source: winner.source,
      sourceRef: winner.sourceRef,
    });
  }

  // Derived edges: a manual edge with the same (from, to) wins field-wise
  // (the derived one only fills missing fields of the first manual one);
  // exact duplicates among derived edges are deduplicated.
  const keptDerived: GraphEdge[] = [];
  for (const edge of derived) {
    const manualIndex = result.findIndex((m) => m.from === edge.from && m.to === edge.to);
    if (manualIndex >= 0) {
      const m = result[manualIndex]!;
      const trigger = m.trigger ?? edge.trigger;
      const label = m.label ?? edge.label;
      const condition = m.condition ?? edge.condition;
      result[manualIndex] = {
        ...(m.id !== undefined ? { id: m.id } : {}),
        from: m.from,
        to: m.to,
        ...(trigger !== undefined ? { trigger } : {}),
        ...(label !== undefined ? { label } : {}),
        ...(condition !== undefined ? { condition } : {}),
        source: m.source,
        sourceRef: m.sourceRef,
      };
      continue;
    }
    const duplicate = keptDerived.some(
      (k) =>
        k.from === edge.from &&
        k.to === edge.to &&
        k.trigger === edge.trigger &&
        k.label === edge.label &&
        k.condition === edge.condition,
    );
    if (!duplicate) keptDerived.push(edge);
  }
  result.push(...keptDerived);

  // Id generation: `e_<from>_<to>`, collisions get suffix _2/_3 in
  // (file, line) order. result is already ordered so that manual edges come
  // before derived ones, each sorted by (file, line).
  const usedIds = new Set<string>();
  for (const edge of result) {
    if (edge.id !== undefined) usedIds.add(edge.id);
  }
  const withIds: GraphEdge[] = [];
  for (const edge of result) {
    if (edge.id !== undefined) {
      withIds.push(edge);
      continue;
    }
    const base = `e_${edge.from}_${edge.to}`;
    let id = base;
    let n = 2;
    while (usedIds.has(id)) {
      id = `${base}_${n}`;
      n++;
    }
    usedIds.add(id);
    withIds.push({ ...edge, id });
  }
  return withIds;
}

/**
 * Merges nodes/edges/flows from all sources. Conflicts between manual
 * sources lead to an [AdapterException] that lists every occurrence.
 */
export function mergeGraph(input: {
  nodes: readonly GraphNode[];
  edges: readonly GraphEdge[];
  flows: readonly GraphFlow[];
}): MergeResult {
  const conflicts: string[] = [];
  const nodes = mergeNodes(input.nodes, conflicts);
  const flows = mergeFlows(input.flows, conflicts);
  const edges = mergeEdges(input.edges, conflicts);
  if (conflicts.length > 0) {
    throw new AdapterException(conflicts);
  }
  return { nodes, edges, flows };
}

File: dart/ductus/lib/src/adapter/merger.dart
/// Internal merge & precedence rules of the adapter.
///
/// `annotation` overrides `derived` field by field; two manual sources with
/// different values for the same field are an error (fail-fast) that cites
/// both source locations.
library;

import 'graph_model.dart';

class MergeResult {
  final List<GraphNode> nodes;
  final List<GraphEdge> edges;
  final List<GraphFlow> flows;

  const MergeResult({
    required this.nodes,
    required this.edges,
    required this.flows,
  });
}

int _byRef(SourceRef a, SourceRef b) {
  final byFile = a.file.compareTo(b.file);
  return byFile != 0 ? byFile : a.line.compareTo(b.line);
}

/// Stable sort by (file, line).
List<T> _sortedByRef<T>(Iterable<T> items, SourceRef Function(T) ref) {
  final indexed = items.toList();
  final order = {for (var i = 0; i < indexed.length; i++) indexed[i]: i};
  indexed.sort((a, b) {
    final c = _byRef(ref(a), ref(b));
    return c != 0 ? c : order[a]!.compareTo(order[b]!);
  });
  return indexed;
}

/// Field-wise merge of a candidate list: first manual value wins, otherwise
/// first derived one; two different manual values ⇒ conflict.
class _FieldMerger<T> {
  final String kind;
  final String id;
  final List<T> manual;
  final List<T> derived;
  final List<String> conflicts;

  _FieldMerger({
    required this.kind,
    required this.id,
    required List<T> candidates,
    required String Function(T) sourceOf,
    required this.conflicts,
  })  : manual = candidates
            .where((c) => sourceOf(c) == SourceKind.annotation)
            .toList(),
        derived = candidates
            .where((c) => sourceOf(c) != SourceKind.annotation)
            .toList();

  V? merge<V>(String field, V? Function(T) get, SourceRef Function(T) ref) {
    V? value;
    T? valueSource;
    for (final c in manual) {
      final v = get(c);
      if (v == null) continue;
      if (value == null) {
        value = v;
        valueSource = c;
      } else if (!_equalValues(value, v)) {
        conflicts.add('Conflict: $kind "$id", field "$field": '
            '"$value" (${ref(valueSource as T)}) vs. "$v" (${ref(c)}).');
      }
    }
    if (value != null) return value;
    for (final c in derived) {
      final v = get(c);
      if (v != null) return v;
    }
    return null;
  }

  static bool _equalValues(Object a, Object b) {
    if (a is List && b is List) {
      if (a.length != b.length) return false;
      for (var i = 0; i < a.length; i++) {
        if (a[i] != b[i]) return false;
      }
      return true;
    }
    return a == b;
  }
}

List<GraphNode> _mergeNodes(List<GraphNode> nodes, List<String> conflicts) {
  final byId = <String, List<GraphNode>>{};
  final idOrder = <String>[];
  for (final node in _sortedByRef(nodes, (n) => n.sourceRef)) {
    byId.putIfAbsent(node.id, () {
      idOrder.add(node.id);
      return [];
    }).add(node);
  }

  final merged = <GraphNode>[];
  for (final id in idOrder) {
    final candidates = byId[id]!;
    if (candidates.length == 1) {
      merged.add(candidates.single);
      continue;
    }
    final m = _FieldMerger<GraphNode>(
      kind: 'Node',
      id: id,
      candidates: candidates,
      sourceOf: (n) => n.source,
      conflicts: conflicts,
    );
    SourceRef ref(GraphNode n) => n.sourceRef;
    final winner = m.manual.isNotEmpty ? m.manual.first : m.derived.first;
    merged.add(GraphNode(
      id: id,
      type: m.merge('type', (n) => n.type, ref) ?? winner.type,
      title: m.merge('title', (n) => n.title, ref),
      flow: m.merge('flow', (n) => n.flow, ref),
      description: m.merge('description', (n) => n.description, ref),
      tags: m.merge('tags', (n) => n.tags.isEmpty ? null : n.tags, ref) ??
          const [],
      source: winner.source,
      sourceRef: winner.sourceRef,
    ));
  }
  return merged;
}

List<GraphFlow> _mergeFlows(List<GraphFlow> flows, List<String> conflicts) {
  final byId = <String, List<GraphFlow>>{};
  final idOrder = <String>[];
  for (final flow in _sortedByRef(flows, (f) => f.sourceRef)) {
    byId.putIfAbsent(flow.id, () {
      idOrder.add(flow.id);
      return [];
    }).add(flow);
  }

  final merged = <GraphFlow>[];
  for (final id in idOrder) {
    final candidates = byId[id]!;
    if (candidates.length == 1) {
      merged.add(candidates.single);
      continue;
    }
    final m = _FieldMerger<GraphFlow>(
      kind: 'Flow',
      id: id,
      candidates: candidates,
      sourceOf: (f) => f.source,
      conflicts: conflicts,
    );
    SourceRef ref(GraphFlow f) => f.sourceRef;
    final winner = m.manual.isNotEmpty ? m.manual.first : m.derived.first;
    merged.add(GraphFlow(
      id: id,
      title: m.merge('title', (f) => f.title, ref),
      start: m.merge('start', (f) => f.start, ref),
      description: m.merge('description', (f) => f.description, ref),
      source: winner.source,
      sourceRef: winner.sourceRef,
    ));
  }
  return merged;
}

List<GraphEdge> _mergeEdges(List<GraphEdge> edges, List<String> conflicts) {
  final sorted = _sortedByRef(edges, (e) => e.sourceRef);
  final manual =
      sorted.where((e) => e.source == SourceKind.annotation).toList();
  final derived =
      sorted.where((e) => e.source != SourceKind.annotation).toList();

  // Manual edges with an explicit id: identity via id, field-wise merge.
  final result = <GraphEdge>[];
  final manualById = <String, List<GraphEdge>>{};
  for (final edge in manual) {
    if (edge.id == null) {
      result.add(edge);
    } else {
      manualById.putIfAbsent(edge.id!, () => []).add(edge);
    }
  }
  for (final entry in manualById.entries) {
    final candidates = entry.value;
    if (candidates.length == 1) {
      result.add(candidates.single);
      continue;
    }
    final m = _FieldMerger<GraphEdge>(
      kind: 'Edge',
      id: entry.key,
      candidates: candidates,
      sourceOf: (e) => e.source,
      conflicts: conflicts,
    );
    SourceRef ref(GraphEdge e) => e.sourceRef;
    final winner = candidates.first;
    result.add(GraphEdge(
      id: entry.key,
      from: m.merge('from', (e) => e.from, ref) ?? winner.from,
      to: m.merge('to', (e) => e.to, ref) ?? winner.to,
      trigger: m.merge('trigger', (e) => e.trigger, ref),
      label: m.merge('label', (e) => e.label, ref),
      condition: m.merge('condition', (e) => e.condition, ref),
      source: winner.source,
      sourceRef: winner.sourceRef,
    ));
  }

  // Derived edges: a manual edge with the same (from, to) wins field-wise
  // (the derived one only fills missing fields of the first manual one);
  // exact duplicates among derived edges are deduplicated.
  final keptDerived = <GraphEdge>[];
  for (final edge in derived) {
    final manualIndex =
        result.indexWhere((m) => m.from == edge.from && m.to == edge.to);
    if (manualIndex >= 0) {
      final m = result[manualIndex];
      result[manualIndex] = GraphEdge(
        id: m.id,
        from: m.from,
        to: m.to,
        trigger: m.trigger ?? edge.trigger,
        label: m.label ?? edge.label,
        condition: m.condition ?? edge.condition,
        source: m.source,
        sourceRef: m.sourceRef,
      );
      continue;
    }
    final duplicate = keptDerived.any((k) =>
        k.from == edge.from &&
        k.to == edge.to &&
        k.trigger == edge.trigger &&
        k.label == edge.label &&
        k.condition == edge.condition);
    if (!duplicate) keptDerived.add(edge);
  }
  result.addAll(keptDerived);

  // Id generation: `e_<from>_<to>`, collisions get a _2/_3 suffix in
  // (file, line) order. result is already ordered so that manual edges come
  // before derived ones, each sorted by (file, line).
  final usedIds = <String>{
    for (final e in result)
      if (e.id != null) e.id!,
  };
  final withIds = <GraphEdge>[];
  for (final edge in result) {
    if (edge.id != null) {
      withIds.add(edge);
      continue;
    }
    final base = 'e_${edge.from}_${edge.to}';
    var id = base;
    var n = 2;
    while (usedIds.contains(id)) {
      id = '${base}_$n';
      n++;
    }
    usedIds.add(id);
    withIds.add(GraphEdge(
      id: id,
      from: edge.from,
      to: edge.to,
      trigger: edge.trigger,
      label: edge.label,
      condition: edge.condition,
      source: edge.source,
      sourceRef: edge.sourceRef,
    ));
  }
  return withIds;
}

/// Merges nodes/edges/flows from all sources. Conflicts between manual
/// sources raise an [AdapterException] listing all source locations.
MergeResult mergeGraph({
  required List<GraphNode> nodes,
  required List<GraphEdge> edges,
  required List<GraphFlow> flows,
}) {
  final conflicts = <String>[];
  final mergedNodes = _mergeNodes(nodes, conflicts);
  final mergedFlows = _mergeFlows(flows, conflicts);
  final mergedEdges = _mergeEdges(edges, conflicts);
  if (conflicts.isNotEmpty) {
    throw AdapterException(conflicts);
  }
  return MergeResult(
    nodes: mergedNodes,
    edges: mergedEdges,
    flows: mergedFlows,
  );
}
